# Анализ сервиса книг

**Цель исследования:** анализ базы данных сервиса для формулировки ценного предложения для нового продукта

# Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

![alt text](https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&table=block&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=2000&userId=&cache=v2)

# Требования

* Посчитайте, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитайте количество обзоров и среднюю оценку;
* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
* Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

# Таблицы

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

**Таблица `books`**

In [2]:
query = (
'''
SELECT *
FROM books
LIMIT(5)
'''
)
pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**Таблица `authors`**

In [3]:
query = (
'''
SELECT *
FROM authors
LIMIT(5)
'''
)
pd.io.sql.read_sql(query, con = engine) 

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**Таблица `ratings`**

In [4]:
query = (
'''
SELECT *
FROM ratings
LIMIT(5)
'''
)
pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**Таблица `reviews`**

In [5]:
query = (
'''
SELECT *
FROM reviews
LIMIT(5)
'''
)
pd.io.sql.read_sql(query, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**Таблица `publishers`**

In [6]:
query = (
'''
SELECT *
FROM publishers
LIMIT(5)
'''
)
pd.io.sql.read_sql(query, con = engine) 

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

## Посчитайте, сколько книг вышло после 1 января 2000 года

In [7]:
query = (
'''
SELECT COUNT(book_id) AS books_total
FROM books
WHERE publication_date >= '2000-01-01';
'''
)
pd.io.sql.read_sql(query, con = engine) 

books_total
0          821

<div style="border-radius: 15px; box-shadow: 4px 4px 4px; border: solid green 2px; padding: 20px">
    
<b>После 1 января 2000 года вышла `821` книга</b>
</div>

## Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [8]:
query = (
'''
SELECT b.book_id,
       b.title,
       COUNT(DISTINCT rv.review_id) AS rewiew_count,
       AVG(rt.rating) AS avg_ratings
FROM books AS b
LEFT JOIN reviews AS rv ON rv.book_id = b.book_id
LEFT JOIN ratings AS rt ON rt.book_id = b.book_id
GROUP BY b.book_id
ORDER BY 3 DESC, 4 DESC;
'''
)
pd.io.sql.read_sql(query, con = engine) 

book_id                                              title  rewiew_count  \
0        948                            Twilight (Twilight  #1)             7   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
2        299  Harry Potter and the Chamber of Secrets (Harry...             6   
3        656                                     The Book Thief             6   
4        734                                   The Glass Castle             6   
..       ...                                                ...           ...   
995      191  Disney's Beauty and the Beast (A Little Golden...             0   
996      387                               Leonardo's Notebooks             0   
997      221                          Essential Tales and Poems             0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
999      808                            The Natural Way to Draw             0   

     avg_ratings  
0       3.662500  
1       4.414634  
2       4.287500  
3       4.264151  
4       4.206897  
..           ...  
995     4.000000  
996     4.000000  
997     4.000000  
998     3.666667  
999     3.000000  

[1000 rows x 4 columns]

<div style="border-radius: 15px; box-shadow: 4px 4px 4px; border: solid green 2px; padding: 20px">
    
<b>Среди ТОП-5 книг по количеству обзоров и среднего рейтинга:</b>

- `Twilight (Twilight #1)`
- `Harry Potter and the Prisoner of Azkaban`
- `Harry Potter and the Chamber of Secrets`
- `The Book Thief`
- `The Glass Castle`
    
Сумерки вышли в топ, лишь потому что у этой книге оказалось больше рецензий

</div>

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [9]:
query = (
'''
SELECT p.publisher,
       COUNT(b.book_id) AS book_total
FROM books AS b
LEFT JOIN publishers AS p ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY book_total DESC
LIMIT(1);
'''
)
pd.io.sql.read_sql(query, con = engine) 

publisher  book_total
0  Penguin Books          42

<div style="border-radius: 15px; box-shadow: 4px 4px 4px; border: solid green 2px; padding: 20px">
    
<b>Издательство, которое выпустило наибольшее число книг толще 50 страниц - `Penguin Books`</b>


</div>

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [10]:
query = (
'''
WITH

books_review AS(
    SELECT b.book_id
    FROM books AS b
    LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
    GROUP BY b.book_id
    HAVING COUNT(DISTINCT rt.rating_id) >= 50
),

ratings_author AS(
    SELECT a.author,
            ROUND(AVG(rt.rating), 2) AS avg_ratings
     FROM books AS b
     JOIN books_review br ON br.book_id = b.book_id
     LEFT JOIN authors AS a ON b.author_id = a.author_id
     LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
     GROUP BY a.author
)

SELECT *
FROM ratings_author
ORDER BY 2 DESC
LIMIT(1);

'''
)
pd.io.sql.read_sql(query, con = engine) 

author  avg_ratings
0  J.K. Rowling/Mary GrandPré         4.29

<div style="border-radius: 15px; box-shadow: 4px 4px 4px; border: solid green 2px; padding: 20px">
    
<b>`J.K. Rowling/Mary GrandPré` является автором книг с самым высоким средним рейтингом книг</b>


</div>

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [11]:
query = (
'''
WITH

users AS(
    SELECT rv.username,
           COUNT(DISTINCT rt.rating_id) AS rating_count
    FROM reviews AS rv
    LEFT JOIN ratings AS rt ON rv.username = rt.username
    GROUP BY rv.username
    HAVING COUNT(DISTINCT rt.rating_id) > 50
),

reviews AS(
    SELECT rv.username,
            COUNT(DISTINCT rv.review_id) AS reviews_count
     FROM reviews rv
     JOIN users AS us ON us.username = rv.username
     GROUP BY rv.username
)

SELECT ROUND(AVG(reviews_count)) AS avg_review
FROM reviews

'''
)
pd.io.sql.read_sql(query, con = engine) 

avg_review
0        24.0

<div style="border-radius: 15px; box-shadow: 4px 4px 4px; border: solid green 2px; padding: 20px">
    
<b>Среднее количество обзоров от пользователей, которые поставили больше 50-ти оценок:</b> `24`


</div>

## Вывод

<div style="border-radius: 15px; box-shadow: 4px 4px 4px; border: solid green 2px; padding: 20px">
    
<b>В ходе анализа требований, выяснили:</b>

- После 1 января 2000 года вышла `821` книга
- Среди ТОП-5 книг по количеству обзоров и среднего рейтинга:
  * `Twilight (Twilight #1)`
  * `Harry Potter and the Prisoner of Azkaban`
  * `Harry Potter and the Chamber of Secrets`
  * `The Book Thief`
  * `The Glass Castle`
- Издательство, которое выпустило наибольшее число книг толще 50 страниц - `Penguin Books`
- `J.K. Rowling/Mary GrandPré` является автором книг с самым высоким средним рейтингом книг
- Среднее количество обзоров от пользователей, которые поставили больше 50-ти оценок: `24`

</div>